### ISSUE TITLE
Improvements for GitHub workflows

### ISSUE THREAD

**Issue by dmarcoux** [rubyforgood/casa#1927]
I've had a look at the GitHub workflows and I've noticed a few possible improvements:
- Avoid running all jobs when linters find issues. It is, just as an example, not needed to run specs in this case since the commit author needs to fix the linter issues first anyway.
- Some jobs share a few if not almost all steps except one. A good example is the `rspec` and `cypress` jobs which are identical, except for the step in which they run `rspec`/`cypress`.  The duplicated steps could be cached once and re-used. This would save some time. Same for other jobs doing operations like setting up Ruby or checking out the code. This could be done once in the first job, then cached and finally restored in the following jobs to avoid repeating the same steps over and over again. This involves creating dependencies between jobs, but it would speed up the CI and remove duplicated code in the GitHub workflows.

Let me know what you think about this. I might be overlooking something. I plan on submitting a PR to address the two issues I listed above.

**Comment by compwron** [comment]
> Avoid running all jobs when linters find issues.
No, we still want to see if tests pass, even if linters fail. It's a matter of developer friendliness. Github actions doesn't charge us, so let's do what's most useful :)

**Comment by compwron** [comment]
caching between rspec and cypress sounds good

**Comment by compwron** [comment]
dependencies between jobs will slow down the overall time to completion since otherwise the jobs run in parallel, but we're not usually in a hurry

**Comment by dmarcoux** [comment]
> > Avoid running all jobs when linters find issues.
> 
> We still want to see if tests pass, even if linters fail. It's a matter of developer friendliness. Github actions doesn't charge us, so let's do what's most useful :)

Alright, is this because it's an open-source project? I'm just wondering.

**Comment by dmarcoux** [comment]
> dependencies between jobs will slow down the overall time to completion since otherwise the jobs run in parallel, but we're not usually in a hurry so that's ok :)

It doesn't necessarily mean that the overall time to completion will be longer, since caching some of the steps might be faster. I'll see what I can do for that.

**Comment by compwron** [comment]
> > > Avoid running all jobs when linters find issues.
> > 
> > 
> > We still want to see if tests pass, even if linters fail. It's a matter of developer friendliness. Github actions doesn't charge us, so let's do what's most useful :)
> 
> Alright, is this because it's an open-source project? I'm just wondering.

That's a good question! I think that developer friendliness is extra important on this project because no one is getting paid to put up with us. I also think that this is a good thing to do at work also, because CI server time is much cheaper than developer time :) 

> It doesn't necessarily mean that the overall time to completion will be longer, since caching some of the steps might be faster. I'll see what I can do for that.

Very cool, thank you for your work! :)

**Comment by dmarcoux** [comment]
Regarding my question `is this because it's an open-source project`, I meant to ask. Is GitHub not charging anything for using GitHub Actions because casa is an open-source project? Or do you have a special agreement with them?

**Comment by compwron** [comment]
Correct 
https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions

> GitHub Actions usage is free for both public repositories and self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account. Any usage beyond the included amounts is controlled by spending limits.
> 
> If you are a monthly-billed customer, your account will have a default spending limit of $0, which prevents additional usage of minutes or storage for private repositories beyond the amounts included with your account. If you pay your account by invoice, your account will have an unlimited default spending limit. For more information, see "About spending limits."
> 
> If you purchased GitHub Enterprise through a Microsoft Enterprise Agreement, you can connect your Azure Subscription ID to your enterprise account to enable and pay for GitHub Actions usage beyond the amounts including with your account. For more information, see "Connecting an Azure subscription to your enterprise."
> 
> Minutes reset every month, while storage usage does not.

**Comment by cam-benfield** [comment]
@dmarcoux, are you working on this or is it open for grabs?

**Comment by dmarcoux** [comment]
It's up for grabs.
